docs/design/direct-assign-volume.md

```diff
@@ -0,0 +1,139 @@
+# Background 
+
+Today, filesystem volumes are passed in to the guest VM utilizing virtio-fs. In a typical flow, a CSI controller
+will make sure backing storage is attached to the node (often a block device), and then a CSI node driver will
+create the filesystem (if needed) and mount it.
+
+![Volume handling without direct-assignment](arch-images/DA-not-enabled.png)
+
+Ideally, for a VM, we will consume the block device itself and not utilize the host mount. This simplifies volume management
+on the host, and is more efficient from a storage IO perspective. 
+
+![Volume handling with direct-assignment](arch-images/DA-UC.png)
+
+# Proposal - Introduce pattern for directly assigning block device for mounting in the guest
+
+In the existing upstream, there is no direct communication between CSI and the container runtime. By the
+time volume information is presented to the runtime, it is a path on the given host filesystem. We need a method to:
+ - enlighten the CSI driver that a given PVC is to be utilized by Kata (or any other runtime that would prefer to mount the volume on their own).
+ - communication path between the CSI node driver and the runtime to detail: what the backing block device is,
+   what kind of filesystem it is, and where it should be mounted with respect to the container rootfs.
+
+Ideally we'd be able to come up with a solution that would facilitate a runtime calling into CSI instead of Kubelet, as is done today for CNI. Working within the existing constraints that exist, we are proposing a pattern that can be used to facilitate direct-volume assignment, while working with other upstream projects for a more direct method. 
+
+## Carrying out the mount
+
+The CSI node driver will need to be enlightened that the volume will be consumed by a VM-based runtime, and will need 
+to communicate with the runtime to describe where the block device lives, what filesystem is on it, and where it
+should be mounted within the container(s).
+
+### Specifiying that the volume will be consumed by a VM-friendly runtime
+
+If a persistent volume [(PV)](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) is intended to be consumed by a VM-based runtime, the user can specify this by adding an annotation to their respect persistent volume claim (PVC). Example:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    skip-hostmount: "true"
```

Comment:
Beware that the SIG storage leads have rejected all attempts so far to pass annotations from PVC to the CSI driver (https://github.com/kubernetes-csi/external-provisioner/issues/86). Expect some opposition when proposing this again...

It can be done by forking the Kubernetes-CSI external-provisioner. But then essentially every CSI driver vendor would have to do that, or we need a second community-maintainer fork of that sidecar.

Reply:
Ack to the hesitation around polluting a CSI plugin with k8 details.

The external-provisioner does provide the PVC name and namespace [today](https://github.com/kubernetes-csi/external-provisioner/blob/1b9d55dd8c7d8c076abeea1039d261723e2d6efa/pkg/controller/controller.go#L113-L114) during CreateVolume. In order to get enlightened, a CSI plugin needs to lookup the PVC annotations from API Server during CreateVolume and store the flag as a field in the CSI `VolumeContext`. This acts as a signal during NodePublish to always skip the mount on the host. 

This approach does "lock" the corresponding PV to be only used by Kata workloads (and not a normal runC workload that would expect the volume to be mounted on the host).